### PR TITLE
GH-33336: [C++][Parquet] Avoid UB on unaligned load

### LINKED ISSUE #33336: [C++] Alignment not enforced; undefined behavior in Parquet writer
It is possible to create arrays using unaligned memory addresses (e.g. for int64). This seems to be in line with the arrow specification which as far as I understand does not require alignment [1].

However, the C++ standard requires alignment, e.g. 8 byte alignment for int64. It is undefined behavior (UB) to create an unaligned pointer / accessing data via an unaligned pointer.

Typically, this is not an issue in practice on x86, since gcc and other compilers mostly emit instructions that can deal with unaligned data. However, for gcc 6.3.0 (and probably up to including gcc versions 7.X), this code:

<https://github.com/apache/arrow/blob/4591d76fce2846a29dac33bf01e9ba0337b118e9/cpp/src/parquet/statistics.cc#L355>

creates an aligned move instruction (movdqa) for the expression `{}values[i]{`}. This, in turn, triggers a SIGSEGV in case `values` is called via an unaligned buffer. Later compiler versions (in particular gcc 9.X used to build the wheels published on pypi) will emit instructions that can deal with unaligned data (movdqu instead of movdqa).

The python script "test1.py" reproduces this issue on python-level; note that it will only trigger a SIGSEGV if compiling arrow with a compiler that emits movdqa for the code linked above, e.g. by using gcc 6.3.0 to compile arrow.

In the wild, unaligned buffers are rare, but can appear, e.g. as a result of deserializing pandas dataframes / numpy arrays using pickle protocol 5 that allows out-of-band byte buffers that are re-used as arrow array buffers.

I think the line to first enter the UB regime is this reinterpret_cast:

https://github.com/apache/arrow/blob/33f2c0ec8e281fc4fe8c03b07ed2d32e343d9b0e/cpp/src/parquet/column_writer.cc#L1592

[1]<https://arrow.apache.org/docs/format/Columnar.html#buffer-alignment-and-padding> merely "recommends" that buffers are aligned, but does not require it.

**Reporter**: [Jochen Ott](https://issues.apache.org/jira/browse/ARROW-18141)
**Assignee**: [Antoine Pitrou](https://issues.apache.org/jira/browse/ARROW-18141) / @pitrou
#### Original Issue Attachments:
- [test1.py](https://issues.apache.org/jira/secure/attachment/13051335/test1.py)
#### PRs and other links:
- [GitHub Pull Request #14488](https://github.com/apache/arrow/pull/14488)

<sub>**Note**: *This issue was originally created as [ARROW-18141](https://issues.apache.org/jira/browse/ARROW-18141). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by pitrou**
Some gcc versions (such as 6.3.0) may emit an aligned-only load instruction, but the Parquet writer can be called with unaligned buffers.
* Closes: #33336

### CODE REVIEW

**Comment by pitrou** [comment]
The fix should probably be made in encoders as well (the various `Put` and `PutSpaced` methods).

**Review by lidavidm — APPROVED** [review]
Two questions:
- Is there a way to unit test this?
- More generally, is there a way we can flush out these issues more systematically? For instance, running tests under an allocator which intentionally returns unaligned buffers?

**Comment by pitrou** [comment]
> More generally, is there a way we can flush out these issues more systematically? For instance, running tests under an allocator which intentionally returns unaligned buffers?

That's a rather good idea. I'll look into this.

**Comment by lidavidm** [comment]
Filed #33933

**Comment by zeroshade** [comment]
The failures here appear to be caused by #33939 i'll wait until that is addressed and we get a clean CI before I merge this.

**Comment by lidavidm** [comment]
@zeroshade time to give this another shot?

**Comment by zeroshade** [comment]
failures don't appear to be related to this change. @lidavidm you agree/objection to me merging?

**Comment by lidavidm** [comment]
I just kicked the jobs to be sure, but otherwise let's merge

**Pull request merged**
